### ISSUE TITLE
SPIRV adding type prefix to constant buffers

### ISSUE THREAD

**Issue by olegviz** [Microsoft/DirectXShaderCompiler#1661]
When generating SPIRV bytecode with -spirv there is a type. prefix added to the constant buffer

In the hlsl shader I have the constant buffer 

cbuffer g_GlobalConstantBufferObject : register (b1)
  {
    float4x4 matMV  : packoffset(c0);   
    float4x4 matMVP : packoffset(c4);   
  }

But in the spirv bytecode I see it became type.g_GlobalConstantBufferObject and 
later in the engine I need to use type_g_GlobalConstantBufferObject to work properly .

when using glslang the type prefix is not added

**Comment by antiagainst** [comment]
We did that because we need to generate both a type and a variable from the above HLSL source snippet. So we need a way to distinguish them, so we named the type with `type.` prefix and keep the variable having the original name.

Are you using the names for reflection? Then you should probably use `-fspv-reflect`: https://khr.io/hlsl2spirv#reflection. Note that according to SPIR-V spec, `Op*Name` should not have semantic effects; we are not guaranteeing back compatibility for it and can change it when needed.

**Comment by DiligentGraphics** [comment]
@antiagainst I am experimenting with the following vertex shader:

```hlsl
cbuffer Constants
{
    float4x4 g_WorldViewProj;
};

struct PSInput 
{ 
    float4 Pos : SV_POSITION; 
    float4 Color : COLOR0; 
};

PSInput main(float3 pos : ATTRIB0, 
             float4 color : ATTRIB1) 
{
    PSInput ps; 
    ps.Pos = mul( float4(pos,1.0), g_WorldViewProj);
    ps.Color = color;
    return ps;
}
```
`glslangValidator.exe -V -H -e main cube.vert.hlsl --hlsl-iomap --auto-map-bindings` produces the following byte code:
```
// Module Version 10000
// Generated by (magic number): 80007
// Id's are bound by 73

                              Capability Shader
               1:             ExtInstImport  "GLSL.std.450"
                              MemoryModel Logical GLSL450
                              EntryPoint Vertex 4  "main" 44 48 57 60
                              Source HLSL 500
                              Name 4  "main"
                              Name 22  "Constants"
                              MemberName 22(Constants) 0  "g_WorldViewProj"
                              Name 24  ""
                              Name 44  "pos"
                              Name 48  "color"
                              Name 57  "@entryPointOutput.Pos"
                              Name 60  "@entryPointOutput.Color"
                              MemberDecorate 22(Constants) 0 RowMajor
                              MemberDecorate 22(Constants) 0 Offset 0
                              MemberDecorate 22(Constants) 0 MatrixStride 16
                              Decorate 22(Constants) Block
                              Decorate 24 DescriptorSet 0
                              Decorate 24 Binding 0
                              Decorate 44(pos) Location 0
                              Decorate 48(color) Location 1
                              Decorate 57(@entryPointOutput.Pos) BuiltIn Position
                              Decorate 60(@entryPointOutput.Color) Location 0
               2:             TypeVoid
               3:             TypeFunction 2
               6:             TypeFloat 32
               7:             TypeVector 6(float) 3
               9:             TypeVector 6(float) 4
              19:             TypeInt 32 1
              20:     19(int) Constant 0
              21:             TypeMatrix 9(fvec4) 4
   22(Constants):             TypeStruct 21
              23:             TypePointer Uniform 22(Constants)
              24:     23(ptr) Variable Uniform
              25:             TypePointer Uniform 21
              29:    6(float) Constant 1065353216
              43:             TypePointer Input 7(fvec3)
         44(pos):     43(ptr) Variable Input
              47:             TypePointer Input 9(fvec4)
       48(color):     47(ptr) Variable Input
              56:             TypePointer Output 9(fvec4)
57(@entryPointOutput.Pos):     56(ptr) Variable Output
60(@entryPointOutput.Color):     56(ptr) Variable Output
         4(main):           2 Function None 3
               5:             Label
              45:    7(fvec3) Load 44(pos)
              49:    9(fvec4) Load 48(color)
              65:     25(ptr) AccessChain 24 20
              66:          21 Load 65
              68:    6(float) CompositeExtract 45 0
              69:    6(float) CompositeExtract 45 1
              70:    6(float) CompositeExtract 45 2
              71:    9(fvec4) CompositeConstruct 68 69 70 29
              72:    9(fvec4) MatrixTimesVector 66 71
                              Store 57(@entryPointOutput.Pos) 72
                              Store 60(@entryPointOutput.Color) 49
                              Return
```
while `dxc.exe -E main -spirv -T vs_6_0 cube.vert.hlsl` produces the following:
```
; SPIR-V
; Version: 1.0
; Generator: Google spiregg; 0
; Bound: 32
; Schema: 0
               OpCapability Shader
               OpMemoryModel Logical GLSL450
               OpEntryPoint Vertex %main "main" %in_var_ATTRIB0 %in_var_ATTRIB1 %gl_Position %out_var_COLOR0
               OpSource HLSL 600
               OpName %type_Constants "type.Constants"
               OpMemberName %type_Constants 0 "g_WorldViewProj"
               OpName %Constants "Constants"
               OpName %main "main"
               OpName %in_var_ATTRIB0 "in.var.ATTRIB0"
               OpName %in_var_ATTRIB1 "in.var.ATTRIB1"
               OpName %out_var_COLOR0 "out.var.COLOR0"
               OpMemberDecorate %type_Constants 0 Offset 0
               OpMemberDecorate %type_Constants 0 MatrixStride 16
               OpMemberDecorate %type_Constants 0 RowMajor
               OpDecorate %type_Constants Block
               OpDecorate %gl_Position BuiltIn Position
               OpDecorate %in_var_ATTRIB0 Location 0
               OpDecorate %in_var_ATTRIB1 Location 1
               OpDecorate %out_var_COLOR0 Location 0
               OpDecorate %Constants DescriptorSet 0
               OpDecorate %Constants Binding 0
      %float = OpTypeFloat 32
    %v4float = OpTypeVector %float 4
%mat4v4float = OpTypeMatrix %v4float 4
%type_Constants = OpTypeStruct %mat4v4float
%_ptr_Uniform_type_Constants = OpTypePointer Uniform %type_Constants
       %void = OpTypeVoid
         %13 = OpTypeFunction %void
    %v3float = OpTypeVector %float 3
%_ptr_Input_v3float = OpTypePointer Input %v3float
%_ptr_Input_v4float = OpTypePointer Input %v4float
%_ptr_Output_v4float = OpTypePointer Output %v4float
    %float_1 = OpConstant %float 1
        %int = OpTypeInt 32 1
      %int_0 = OpConstant %int 0
%_ptr_Uniform_mat4v4float = OpTypePointer Uniform %mat4v4float
  %Constants = OpVariable %_ptr_Uniform_type_Constants Uniform
%in_var_ATTRIB0 = OpVariable %_ptr_Input_v3float Input
%in_var_ATTRIB1 = OpVariable %_ptr_Input_v4float Input
%gl_Position = OpVariable %_ptr_Output_v4float Output
%out_var_COLOR0 = OpVariable %_ptr_Output_v4float Output
       %main = OpFunction %void None %13
         %22 = OpLabel
         %23 = OpLoad %v3float %in_var_ATTRIB0
         %24 = OpLoad %v4float %in_var_ATTRIB1
         %25 = OpCompositeExtract %float %23 0
         %26 = OpCompositeExtract %float %23 1
         %27 = OpCompositeExtract %float %23 2
         %28 = OpCompositeConstruct %v4float %25 %26 %27 %float_1
         %29 = OpAccessChain %_ptr_Uniform_mat4v4float %Constants %int_0
         %30 = OpLoad %mat4v4float %29
         %31 = OpMatrixTimesVector %v4float %30 %28
               OpStore %gl_Position %31
               OpStore %out_var_COLOR0 %24
               OpReturn
               OpFunctionEnd
```
I then use SPIRV-Cross to get reflection information from SPIRV-binary. For glslang byte code, it enumerates uniform buffer name as `Constants` which is what I would expect. However for dxc, it enumerates it as `type_Constants`. I am trying to understand if this is expected behavior or should SPIRV-Cross return the name of the variable rather than the name of the type in dxc case?

**Comment by antiagainst** [comment]
Interestingly, we used to prefix the variable generated from cbuffer/tbuffer with `var.` and then we heard complains from developers that it didn't play nicely with their reflection pipeline, so we dropped it (https://github.com/Microsoft/DirectXShaderCompiler/pull/1071). 

The lesson learned from it is that instead of relying on **every** compiler to have the same mechanism for emitting names, which is spec'd as having no semantic impact and can be freely deleted, we should have a more reliable and standardized mechanism. That's why we proposed an [HLSL functionality extension](https://github.com/KhronosGroup/SPIRV-Registry/blob/master/extensions/GOOGLE/SPV_GOOGLE_hlsl_functionality1.asciidoc). Both Glslang (via `-fhlsl_functionality1` IIRC) and DXC (via `-fspv-reflect`) can emit it. I'm not clear about SPIRV-Cross; but I think it should be able to digest the info emitted with the above command-line options? The functionality has been there in both compilers for quite some time already. And [SPIRV-Reflect](https://github.com/chaoticbob/SPIRV-Reflect) works well with it.
